Support setting target environment version, e.g. Vulkan 1.1

dneto0 opened the pull request:
NOTE: We no longer can compile OpenGL compatibility profile shaders to SPIR-V.
These will error out.

C and C++ API: Set nontrivial target environment version

Glslc --target-env accepts optional version suffix, e.g. vulkan1.1

dneto0 commented:
FYI:  CI-ubuntu-clang-debug failure is a config failure: "    clang  is not a full path and was not found in the PATH."

dneto0 commented:
PTAL @antiagainst .

 @dgkoch has asked for this to land sooner rather than later.

antiagainst commented on libshaderc_util/src/compiler.cc:
```diff
@@ -139,6 +186,25 @@ std::tuple<bool, std::vector<uint32_t>, size_t> Compiler::Compile(
   std::vector<uint32_t>& compilation_output_data = std::get<1>(result_tuple);
   size_t& compilation_output_data_size_in_bytes = std::get<2>(result_tuple);
 
+  // Check target envirohnment.
```
s/envirohnment/environment/

dneto0 replied:
Yes.

dgkoch commented on libshaderc_util/include/libshaderc_util/compiler.h:
```diff
@@ -103,11 +106,20 @@ class Compiler {
     OpenGLCompat,
   };
 
+  // Target environment versions.  These bnumbgers match those used by Glslang.
```
"bnumbgers" typo?

dgkoch commented on libshaderc_util/include/libshaderc_util/compiler.h:
```diff
@@ -103,11 +106,20 @@ class Compiler {
 
+  // Target environment versions.  These bnumbgers match those used by Glslang.
+  enum class TargetEnvVersion : uint32_t {
+    // For Vulkan, use numbering schmee from vulkan.h
```
"schmee"?

dgkoch commented on libshaderc_util/include/libshaderc_util/compiler.h:
```diff
@@ -103,11 +106,20 @@ class Compiler {
+  enum class TargetEnvVersion : uint32_t {
+    // For Vulkan, use numbering schmee from vulkan.h
+    Vulkan_1_0 = ((1 << 22)),              // Default to Vuilkan 1.0
+    Vulkan_1_1 = ((1 << 22) | (1 << 12)),  // Default to Vuilkan 1.0
```
Vuilkan

dgkoch commented on libshaderc_util/include/libshaderc_util/compiler.h:
```diff
@@ -103,11 +106,20 @@ class Compiler {
+    // For Vulkan, use numbering schmee from vulkan.h
+    Vulkan_1_0 = ((1 << 22)),              // Default to Vuilkan 1.0
+    Vulkan_1_1 = ((1 << 22) | (1 << 12)),  // Default to Vuilkan 1.0
+    // For Op[enGL, use the numbering from #version in shaders.
```
Op[enGL

dgkoch commented on libshaderc_util/include/libshaderc_util/compiler.h:
```diff
@@ -103,11 +106,20 @@ class Compiler {
   enum class TargetEnv {
-    Vulkan,
-    OpenGL,
+    Vulkan,  // Default to Vuilkan 1.0
```
Vuilkan

dneto0 replied:
oh my.  I must have been tired.  Actually I was on a plane and comments were dark blue on black... :-(

antiagainst commented:
Also, we should probably update the glslc manual about the changes.

antiagainst commented on glslc/src/main.cc:
```diff
@@ -135,9 +135,14 @@ An input file of - represents standard input.
+                    the client version.  Values are:
+                        vulkan1.0       # The default
+                        vulkan1.1
+                        vulkan          # Same as vulkan1.0
```
A question  about the `vulkan` parameter suddenly comes to me.

What's our guarantee of this parameter? Are we gonna change what it aliases to when, say, Vulkan 1.2 or 1.N releases?

If not, that effectively makes it just a way of saying `vulkan1.0`, which would be handy for now but in the future developers will mostly want vulkan 1.N. So it will likely be less useful.

If we would like to change it when newer version of Vulkan come out, then how we decide which version to alias to since we are not alias to the newest version here?

Also note that when we have a `python` or other binary or library without version number in the system, they alias to the newest ones. If  you want to the older ones, you need to specify `python2.6` or something. I'm not sure whether that's particularly related here, but seems it's a convention?

antiagainst replied:
Okay, I think maybe make `vulkan` alias to the previous major release is a good balance between new functionality and driver stability?

dneto0 replied:
JohnK and I independently came to the same conclusion that at least for now defaulting to Vulkan 1.0 is a sane default.  It also happens to retain backward compatibility.
It is a judgement call though.

antiagainst replied:
That's fair.

dneto0 commented on glslc/test/option_target_env.py:
```diff
@@ -80,13 +87,60 @@ class TestDefaultTargetEnvWithVulkanShader(expect.ValidObjectFile):
+
+
+@inside_glslc_testsuite('OptionTargetEnv')
+class TestTargetEnvEqVulkan1_0WithVulkan1_1ShaderFails(expect.ErrorMessageSubstr):
```
Yes!  Deleting the one at line 98.

dneto0 commented on glslc/test/option_target_env.py:
```diff
@@ -80,13 +87,60 @@ class TestDefaultTargetEnvWithVulkanShader(expect.ValidObjectFile):
+@inside_glslc_testsuite('OptionTargetEnv')
+class TestTargetEnvEqVulkan1_1WithVulkan1_0ShaderSucceeds(expect.ValidObjectFile):
+    shader = FileShader(vulkan_vertex_shader(), '.vert')
+    glslc_args = ['--target-env=vulkan1.0', '-c', shader]
```
Yes!

dneto0 commented on libshaderc_util/include/libshaderc_util/compiler.h:
```diff
@@ -214,8 +227,9 @@ class Compiler {
-  // Sets the target environment.
-  void SetTargetEnv(TargetEnv env);
+  // Sets the target environment, including version.
+  // 0 means the default version for the
```
yes.  whoops.

dneto0 commented:
Thanks for the reviews.  I've addressed all feedback, and also updated the comments at shaderc_compile_set_target_env and its friends.

@antiagainst PTAL

antiagainst commented:
Appveyor & Travis is fine. I'll ignore Google bots for now.

ehsannas commented:
The Google bot failures are new and unrelated to this CL. I will work on fixing them.
